compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/IncrementalCompilerRunner.kt

```diff
@@ -51,14 +52,24 @@ abstract class IncrementalCompilerRunner<
     cacheDirName: String,
     protected val reporter: BuildReporter,
     protected val buildHistoryFile: File,
-    // there might be some additional output directories (e.g. for generated java in kapt)
-    // to remove them correctly on rebuild, we pass them as additional argument
-    private val additionalOutputFiles: Collection<File> = emptyList(),
+
+    /**
+     * Output directories of the compilation. These include:
+     *   1. The classes output directory
+     *   2. [workingDir]
+     *   3. Any additional output directories (e.g., classpath snapshot directory or Kapt generated-stubs directory)
```

Comment:
Great catch! 

I find it very confusing that classpath snapshot is not cleaned together with the other outputs, and that was only to support an unimportant use case*, so I decided to change that.
 
\* The use case is that incremental compilation is requested (`InputChanges.incremental == true`), but non-incremental compilation is performed and fails. Previously we wouldn't delete the classpath snapshot so that the next build could be incremental. But because non-incremental compilation was already performed in this build, I think it is not important to make the next build incremental.

Also, the fact that we can't compile incrementally means there is a bug in the tools, and once we've fixed that, this use case will no longer be possible.

Reply:
kk, that sounds good. This ensures that we don't need to back up class snapshot on a "normal" syntax error, as we'll have an execution flow that does not throw i.e. we'll not write a new version of the classpath snapshots => the next run will still be incremental.